crates/trie/sparse/src/trie.rs

```diff
@@ -257,8 +261,244 @@ impl RevealedSparseTrie {
     }
 
     /// Remove leaf node from the trie.
-    pub fn remove_leaf(&mut self, _path: Nibbles) {
-        unimplemented!()
+    pub fn remove_leaf(&mut self, path: Nibbles) -> SparseTrieResult<()> {
+        self.prefix_set.insert(path.clone());
+        let existing = self.values.remove(&path);
+        if existing.is_none() {
+            // trie structure unchanged, return immediately
+            return Ok(())
+        }
+
+        let mut removed_nodes = self.take_nodes_for_path(&path)?;
+        debug!(target: "trie::sparse", ?path, ?removed_nodes, "Removed nodes for path");
+        // Pop the first node from the stack which is the leaf node we want to remove.
+        let Some(mut child) = removed_nodes.pop_back() else { return Ok(()) };
```

Comment:
note: `else` case should never happen here because revealed trie must at least contain root node at all times